### PR TITLE
[SAM] Makes S3 Bucket parameter optional and creates bucket automatically in the deployment region

### PR DESCRIPTION
**Pull request by heitorlessa**
Hi everyone,

 This PR makes it possible to omit `--s3-bucket` and AWS CLI will create a S3 bucket in the deployment region specified - This makes the following possible:

* Ability to launch the same SAM template in multiple regions
* Error out earlier if S3 Bucket and Deployment region differs instead of waiting for Cloudformation to be deployed and fail
* Ability to automatically create S3 Bucket if no `--s3-bucket` was passed

 This is a non-breaking change and those who are currently using `--s3-bucket` should continue to work just fine.

**Usage: Deploy SAM stack in Ireland region without specifying the bucket**

```bash
aws cloudformation package \
    --template-file template.yaml \
    --output-template-file packaged.yaml \
    --region eu-west-1 

aws cloudformation deploy \
    --template-file template.yaml \
    --stack-name test-sam \
    --region eu-west-1 \
    --capabilities CAPABILITY_IAM 
```

> **This will create a bucket named `sam-<region>-<accountId>`**

cc'ing contributors @sthulb @igngar

### CODE REVIEW

**Comment by heitorlessa** [comment]
cc @sanathkr @jamesls

File: awscli/customizations/cloudformation/package.py [hunk 1]
```diff
@@ -112,19 +115,70 @@ class PackageCommand(BasicCommand):
...
+                        self.MSG_PACKAGE_S3_BUCKET_CREATION.format(
+                            bucket=bucket, region=region))
+
+                    _s3_params = {
+                        "all_regions": {
+                            "Bucket": bucket
+                        },
+                        "us_standard": {
+                            "Bucket": bucket,
+                            "CreateBucketConfiguration": {
```
Comment (sanathkr) [inline awscli/customizations/cloudformation/package.py]: I don't understand. I thought the location constraint was for non-us-standard regions. Am I missing something?

**Comment by sthulb** [comment]
I think adding the new flag would be redundant. There’s already a dry run flag to perform no-ops. 

- thulsimo@

File: awscli/customizations/cloudformation/package.py [hunk 2]
```diff
@@ -112,19 +115,70 @@ class PackageCommand(BasicCommand):
         }
     ]
 
+    def _get_bucket_region(self, s3_bucket, s3_client):
+        s3_loc = s3_client.get_bucket_location(
+            Bucket=s3_bucket).get("LocationConstraint", "us-east-1")
+        return s3_loc.get
```
[the remark this answers is not in the document]
Comment (sthulb) [reply awscli/customizations/cloudformation/package.py]: It allows the function to be mocked for the tests easier.

File: awscli/customizations/cloudformation/package.py [hunk 3]
```diff
@@ -112,19 +115,70 @@ class PackageCommand(BasicCommand):
...
+
     def _run_main(self, parsed_args, parsed_globals):
+        region = parsed_globals.region if parsed_globals.region else "us-east-1"
```
[the remark this answers is not in the document]
Comment (sthulb) [reply awscli/customizations/cloudformation/package.py]: That’s a good idea.

**Comment by heitorlessa** [comment]
Thanks for all comments @sanathkr - We'll go through them by max EOW. 

We discussed having that option on s3 bucket name and after considering all scenarios and trying in practice we found that it only adds more complexity -- Main idea was to make developers to remember less parameters and get going more quickly. 

The ideal scenario we thought was for SAM to support a "project_name" variable of some sort, of which it'd allow the developer to have the S3 bucket automatically created and named using the project_name + uuid. 

Happy to hear your thoughts on this though

File: awscli/customizations/cloudformation/package.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (heitorlessa) [reply awscli/customizations/cloudformation/package.py]: That was our initial surprise when hacking that during the hackathon - If you don't supply "--region" during the package command "parsed_globals.region" will be "None" and we were expecting to be whatever was set in the AWS CLI - **That's the reason we had this one-liner if**.

What's set in the AWS CLI like regions for a profile (default, lab, etc.) only seems to work when you initiate a connection with a service (self._session <- contains a dict of regions set in a profile and likely use those in the absence of one). 

Given that we need the region set as a parameter to run some additional logic we have two options here:

1) Error out if not supplied and end execution there as simple as that
2) Try capture the default region configured in the AWS CLI (there's gotta be an easy way other than parsing self._session from 'default' or from a 'profile' if set)

Implemented 1st option for now and commit to follow

Thoughts?

File: awscli/customizations/cloudformation/package.py (hunk 1 above)
Comment (heitorlessa) [reply awscli/customizations/cloudformation/package.py]: You're right - Fixed.

**Comment by heitorlessa** [comment]
As soon as we agree on the `region` I'll write an additional unit test to take that into account - Let me know what you think @sanathkr

**Comment by sanathkr** [comment]
@heitorlessa Can you fix the test failures?

**Comment by heitorlessa** [comment]
sorry @sanathkr but which failures? I ran them before pushing the latest commits and the only one that is complaining now is a health check from Codecov that I don't know how to improve - If that's the case, any suggestions that I can go and fix?

```python
20:27:59 ➜ aws-cli git:(sam-package-enhanced) venv:(aws-cli-hackathon) python:(🐍 aws-cli-hackathon) python -m unittest -v tests/unit/customizations/cloudformation/test_package.py
test_main (tests.unit.customizations.cloudformation.test_package.TestPackageCommand) ...
Successfully packaged artifacts and wrote output template to file ./oputput.
Execute the following command to deploy the packaged template
aws cloudformation deploy --template-file /Users/lessa/new-packaging-hackathon/aws-cli/oputput --stack-name <YOUR STACK NAME>
/Users/lessa/new-packaging-hackathon/aws-cli/tests/unit/customizations/cloudformation/test_package.py:81: DeprecationWarning: Please use assertEqual instead.
  self.assertEquals(rc, 0)

Successfully packaged artifacts and wrote output template to file ./oputput.
Execute the following command to deploy the packaged template
aws cloudformation deploy --template-file /Users/lessa/new-packaging-hackathon/aws-cli/oputput --stack-name <YOUR STACK NAME>
ok
test_main_bucket_different_deployment_region (tests.unit.customizations.cloudformation.test_package.TestPackageCommand) ... ok
test_main_empty_region_error (tests.unit.customizations.cloudformation.test_package.TestPackageCommand) ... ok
test_main_error (tests.unit.customizations.cloudformation.test_package.TestPackageCommand) ... ok
test_main_without_bucket (tests.unit.customizations.cloudformation.test_package.TestPackageCommand) ...
Successfully packaged artifacts and wrote output template to file ./oputput.
Execute the following command to deploy the packaged template
aws cloudformation deploy --template-file /Users/lessa/new-packaging-hackathon/aws-cli/oputput --stack-name <YOUR STACK NAME>

Successfully packaged artifacts and wrote output template to file ./oputput.
Execute the following command to deploy the packaged template
aws cloudformation deploy --template-file /Users/lessa/new-packaging-hackathon/aws-cli/oputput --stack-name <YOUR STACK NAME>
ok
test_write_output_to_stdout (tests.unit.customizations.cloudformation.test_package.TestPackageCommand) ... ok

----------------------------------------------------------------------
Ran 6 tests in 0.014s

OK
```

**Comment by sanathkr** [comment]
This: https://codecov.io/gh/aws/aws-cli/compare/809ddea13fabc2be61634bbb8242fec62e5b07fa...8403b76eaf7b78c113577c5705788d617b63f030

AFAIK, I think it is complaining about code coverage drop..

**Comment by heitorlessa** [comment]
@sanathkr Did some refactor in the original tests including mine and reached 100% according to [`coverage`](https://coverage.readthedocs.io) - Waiting for Codecov to kick in within a few minutes.

There is one conflict on `exceptions.py` as a new commit added a new Exception which is super simple to resolve, but although I tried resolving (commit 37d1e22) it it seems I can't due to not having write access.

**Comment by Dunedan** [comment]
@sanathkr Any update on this? I'd really love to see this getting merged.

**Comment by RyanFitzSimmonsAK** [comment]
Hi @heitorlessa, thanks for creating this PR and for your patience. Could you please rebase this PR against the latest branch for the team’s further consideration going forward?

**Comment by RyanFitzSimmonsAK** [comment]
Going to close this PR for the time being.  If someone later decides they want to pick this up, rebase it and it will be reopened for consideration.

**Pull request closed without merging**